File: Problems/387_First_Unique_Character_in_a_String.py
# another solution using built-in function, it's kinda like cheating, I suggest using my second solution
from collections import Counter, defaultdict
class Solution:
    def firstUniqChar(self, s: str) -> int:
        counter = Counter(s)
        for i, c in enumerate(s):
            if counter[c] == 1:
                return i
        return -1

# using defaultdict makes it easy
class Solution:
    def firstUniqChar(self, s: str) -> int:
        hm = defaultdict(int)
        for c in s:
            hm[c] += 1
        
        for i, c in enumerate(s):
            if hm[c] == 1:
                return i
        return -1

class Solution:
    def firstUniqChar(self, s: str) -> int:
        hash_map = {}
        for c in s:
            if c not in hash_map:
                hash_map[c] = 1
            else:
                hash_map[c] += 1

        for i, c in enumerate(s):
            if hash_map[c] == 1:
                return i
        return -1

File: Problems/387_First_Unique_Character_in_a_String.cpp
class Solution {
public:
    int firstUniqChar(string s) {
        unordered_map<char, int> hm;
        for (auto c : s) {
            ++hm[c];
        }
        for (int i = 0; i < s.length(); ++i) {
            char c = s[i];
            if (hm[c] == 1) return i;
        }
        return -1;
    }
};